### PR TITLE
【PerfXLab】add group_gemm kernel

### PR DESCRIPTION
**Pull request by v2yield**
### PR Category
[ Operator ]

### Type of Change
[ Performance Optimization]

### Description
add group gemm kernel for moe

### Performance
group_gemm.py::test_group_gemm_speedup[dtype0-16-512-2048]                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
Base Latency (ms)    Gems Latency (ms)         Gems Speedup               TFLOPS                                                                                                                                                                              
--------------------------------------------------------------------------------                                                                                                                                                                              
       3.519040             0.196064               17.948               42.785                                                                                                                                                                                
                                                                                                                                                                                                                                                              
PASSED
                                                                                                                                                                                                                                                        
group_gemm.py::test_group_gemm_speedup[dtype0-16-2560-2048]                                                                                                                                                                                                   
                                                                                                                                                                                                                                                              
Base Latency (ms)    Gems Latency (ms)         Gems Speedup               TFLOPS                                                                                                                                                                              
--------------------------------------------------------------------------------                                                                                                                                                                              
       3.515488             0.287200               12.241              145.165                                                                                                                                                                                
                                                                                                                                                                                                                                                              
PASSED 
                                                                                                                                                                                                                                                       
group_gemm.py::test_group_gemm_speedup[dtype0-64-2048-128]                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
Base Latency (ms)    Gems Latency (ms)         Gems Speedup               TFLOPS                                                                                                                                                                              
--------------------------------------------------------------------------------                                                                                                                                                                              
       6.098080             0.511776               11.916               11.390                                                                                                                                                                                
                                                                                                                                                                                                                                                              
PASSED

### CODE REVIEW

File: src/flag_gems/ops/group_gemm.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+@pytest.mark.parametrize(
+    "groups, N, K", [(16, 512, 2048), (16, 2560, 2048), (64, 2048, 128)]
+)
+@pytest.mark.parametrize("dtype", [torch.bfloat16])
+def test_group_gemm_speedup(groups, N, K, dtype):
+    M = [
+        1,
+        2,
+        4,
+        8,
```
Comment (tengqm) [inline src/flag_gems/ops/group_gemm.py]: Is it okay to write arrays like this in the following way?

```python

M = [
    1, 2, 4, 8, 16, 24, 32, 40,
    48, 56, 64, 72, 80, 88, 96, 104,
    112, 120,
]
```

We don't have to write one number per line, right?

File: src/flag_gems/ops/group_gemm.py [hunk 2]
```diff
@@ -0,0 +1,781 @@
...
+        M,
+        N,
+        K,
+        C.stride(0),
+        C.stride(1),
+    )
+    return C
+
+
+@pytest.mark.parametrize(
```
Comment (tengqm) [inline src/flag_gems/ops/group_gemm.py]: We don't write test cases in the same file.

File: src/flag_gems/ops/group_gemm.py [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+    print(header_break)
+    print(
+        f"{latency_base_str:>15}",
+        f"{latency_str:>20}",
+        f"{speedup_str:>20}",
+        f"{tflops_str:>20}\n",
+    )
+
+
+@pytest.mark.parametrize(
```
Comment (tengqm) [inline src/flag_gems/ops/group_gemm.py]: and, we separate performance tests from accuracy/correctness tests.

File: src/flag_gems/ops/group_gemm.py [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+    offs = torch.tensor(
+        [sum(M_list[: i + 1]) for i in range(group_size)],
+        dtype=torch.int32,
+        device="cuda",
+    )
+
+    ref = torch._grouped_mm(mat_a, mat_b, offs)
+    res = group_mm(mat_a, mat_b, offs)
+    print(ref)
+    print(res)
```
Comment (tengqm) [inline src/flag_gems/ops/group_gemm.py]: Don't print anything in your unit tests.

File: src/flag_gems/ops/__init__.py [hunk 5]
```diff
@@ -355,6 +356,7 @@
     "get_scheduler_metadata",
     "glu",
     "glu_backward",
+    "group_mm",
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py:483]: I'm not sure if this has to be registered at `src/flag_gems/__init__.py`.
Please double check.

Comment (v2yield) [reply src/flag_gems/ops/__init__.py:483]: If this line is removed, it will report no attribute 'group_mm' found on the module 'flag_gems'.

Comment (tengqm) [reply src/flag_gems/ops/__init__.py:483]: I meant `src/flag_gems/__init__.py`, in addition to `src/flag_gems/ops/__init__.py`.

File: src/flag_gems/ops/group_gemm.py [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+    assert B.dim() == 3
+    M, K = A.shape
+
+    num_groups, BK, N = B.shape
+    strideBK, strideBN = B.stride(1), B.stride(2)
+
+    assert num_groups == offs.numel()
+    NUM_SMS = torch.cuda.get_device_properties("cuda").multi_processor_count
+    C = A.new_empty(M, N)
+    if hasattr(triton.tools.tensor_descriptor, "TensorDescriptor") and supports_tma():
```
Comment (Ea760) [inline src/flag_gems/ops/group_gemm.py]: grouped_gemm_tma_kernel might call tl.make_tensor_descriptor during compilation, so CICD fails on Triton versions without this api, try to use xxx = hasattr(tl, "make_tensor_descriptor") and set it to none when the check fails.

Comment (v2yield) [reply src/flag_gems/ops/group_gemm.py]: Sorry, I didn't quite understand what you meant. 
Should it be: `if hasattr(triton.tools.tensor_descriptor, "TensorDescriptor") and supports_tma():` changed to `if hasattr(tl, "make_tensor_descriptor") and hasattr(triton.tools.tensor_descriptor, "TensorDescriptor") and supports_tma():`?

File: src/flag_gems/ops/__init__.py [hunk 7]
```diff
@@ -468,6 +469,7 @@
...
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py]: pls keep the list order; and
check if `group_mm` should be registered to `src/flag_gems/__init__.py` as well.

**Review by tengqm — APPROVED** [review]

**Comment by huangyiqun** [comment]
Please resolve conflicts.

**Review by huangyiqun — APPROVED** [review]
